# Data Challenge - Πρόβλεψη Πληρότητας Πτήσεων

Στα πλαίσια της εργασίας του μαθήματος "Εξόρυξη Γνώσης από Βάσεις Δεδομένων και τον Παγκόσμιο Ιστό", θα δουλέψετε πάνω σε ένα πρόβλημα κατηγοριοποίησης. Συγκεκριμένα, σας δίνεται ένα σύνολο δεδομένων το οποίο αποτελείται από μερικές χιλιάδες πτήσεις, όπου κάθε πτήση περιγράφεται απο ένα σύνολο μεταβλητών (αεροδρόμιο αναχώρησης, αεροδρόμιο άφιξης, κτλ). Κάθε πτήση χαρακτηρίζεται επίσης από μια μεταβλητή που σχετίζεται με τον αριθμό των επιβατών της πτήσης (π.χ. κάθε τιμή της μεταβλητής σχετίζεται με ενα εύρος πλήθους επιβατών). Για κάποιες πτήσεις, η τιμή της μεταβλητής  είναι γνωστή, ενώ για άλλες όχι. Στόχος σας είναι να προβλέψετε την τιμή της μεταβλητής για τις πτήσεις για τις οποίες δεν είναι διαθέσιμη.

### Σύνολο Δεδομένων

Το αρχείο με όνομα `train.csv` περιέχει τα δεδομένα εκπαίδευσης (training set) του προβλήματος, ενώ το αρχείο `test.csv` περιέχει τα δεδομένα ελέγχου (test set) του προβλήματος. Κάθε γραμμή των δυο αυτών αρχείων αντιστοιχεί σε μια πτήση η οποία χαρακτηρίζεται από τις εξής μεταβλητές:

Μεταβλητή | Περιγραφή
--- | --- 
DateOfDeparture | Ημερομηνία αναχώρησης
Departure | Κωδικός αεροδρομίου αναχώρησης
CityDeparture | Πόλη αναχώρησης
LongitudeDeparture 	 | Γεωγραφικό μήκος αεροδρομίου αναχώρησης
LatitudeDeparture 	 | Γεωγραφικό πλάτος αεροδρομίου αναχώρησης
Arrival | Κωδικός αεροδρομίου άφιξης
CityArrival | Πόλη άφιξης
LongitudeArrival | Γεωγραφικό μήκος αεροδρομίου άφιξης
LatitudeArrival | Γεωγραφικό πλάτος αεροδρομίου άφιξης
WeeksToDeparture | Πόσες εβδομάδες πριν την αναχώρηση της πτήσης κατά μέσο όρο έκλεισαν οι επιβάτες τα εισητήριά τους
std_wtd | Τυπική απόκλιση για το παραπάνω 
Το training set περιέχει μια επιπλέον μεταβλητή (`PAX`) η οποία έχει σχέση με τον αριθμό των επιβατών της πτήσης. Η μεταβλητή αυτή παίρνει 8 διαφορετικές τιμές (τιμές από 0 έως 7 οπότε 8 κατηγορίες συνολικά). Κάθε κατηγορία υποδηλώνει πόσοι περίπου επιβάτες χρησιμοποίησαν την πτήση. Οι αριθμοί στις κατηγορίες έχουν ανατεθεί με τυχαίο τρόπο, οπότε μην θεωρήσετε ότι η κατηγορία 0 υποδηλώνει πολύ λίγους επιβάτες ενώ η κατηγορία 7 πάρα πολλούς επιβάτες. Η μεταβλητή `PAX` λείπει από το test set καθώς πρόκειται για την μεταβλητή που πρέπει να προβλέψετε στα πλαίσια της παρούσας εργασίας.

Παρακάτω σας δίνεται κώδικας ο οποίος φορτώνει τα δεδομένα εκπαίδευσης σε ένα DataFrame της βιβλιοθήκης Pandas και τυπώνει τις πρώτες 5 γραμμές. Οπότε μπορείτε να δείτε τις 12 μεταβλητές του προβλήματος.

In [1]:
import pandas as pd

df_train = pd.read_csv('dataset/train.csv')
df_train.head()


DateOfDeparture Departure  CityDeparture  LongitudeDeparture  \
0      2011-12-05       ORD        Chicago           41.978603   
1      2013-03-01       ATL        Atlanta           33.636719   
2      2012-02-06       ORD        Chicago           41.978603   
3      2012-12-10       ORD        Chicago           41.978603   
4      2012-04-13       SFO  San Francisco           37.618972   

   LatitudeDeparture Arrival    CityArrival  LongitudeArrival  \
0         -87.904842     EWR         Newark         40.692500   
1         -84.428067     LGA       New York         40.777245   
2         -87.904842     BOS         Boston         42.364347   
3         -87.904842     SFO  San Francisco         37.618972   
4        -122.374889     JFK       New York         40.639751   

   LatitudeArrival  WeeksToDeparture    std_wtd  PAX  
0       -74.168667          8.352941   5.667243    7  
1       -73.872608         10.421053  10.001754    7  
2       -71.005181          9.250000   7.136821    7  
3      -122.374889          8.666667   7.404291    7  
4       -73.778925         14.037037   9.858544    7

Επίσης, όσον αφορά τις διαστάσεις του training set, όπως βλέπετε παρακάτω, αποτελείται από 8899 γραμμές και 12 στήλες.

In [2]:
df_train.shape

(8899, 12)

In [3]:
import dateutil.parser as p
import numpy as np

dates = df_train[['DateOfDeparture']]

datetime=[p.parse(dates.iloc[i]['DateOfDeparture']) for i in range(dates.shape[0])]

date_number=[datetime[i].isoweekday() for i in range(dates.shape[0])]

date_month=[datetime[i].isocalendar()[1] for i in range(dates.shape[0])]

date_number=pd.DataFrame(date_number,columns=['Date'])
date_month=pd.DataFrame(date_month,columns=['Week'])

df_train['Date']=date_number.values
df_train['Week']=date_month.values


score = np.zeros((dates.shape[0]),dtype=int)
for i in range(dates.shape[0]):
    if datetime[i].isoweekday()==6 or datetime[i].isoweekday==7:
        score[i]=1
    
scor=pd.DataFrame(score,columns=['Weekend'])
df_train['Weekend']=scor.values


score = np.empty((dates.shape[0]),dtype=str)
for i in range(dates.shape[0]):
    if datetime[i].isoweekday() == 4 and 22 <= datetime[i].day < 29 and datetime[i].month==11:
        score[i]='Independance Day'
    if 24 <= datetime[i].day <= 25 and datetime[i].month==12:
        score[i]='Christmas'
    if (datetime[i].day == 31 and datetime[i].month==12) or (datetime[i].day == 1 and datetime[i].month==1):
        score[i]='New Year'
    
scor=pd.DataFrame(score,columns=['Celebs'])
df_train['Celebs']=scor.values

df_train[['WeeksToDeparture']]= pd.cut(df_train['WeeksToDeparture'],  [0,3.6,4.01,6,8,9,10,11,12,13,14,15,16,17,18.3,20.5,24],labels=[ '0-3', '3-4','4-6','6-8','8-9','9-10','10-11','11-12','12-13','13-14','14-15','15-16','16-17','17-18','18-20','20-24'])

cols = df_train.columns.tolist()
cols =cols[-4:]+cols[:-4]
df_train=df_train[cols]

In [4]:
import pandas as pd
from sklearn import preprocessing
# Create a minimum and maximum processor object
min_max_scaler = preprocessing.MinMaxScaler()

# Create an object to transform the data to fit minmax processor

df_train[['LatitudeDeparture']]= min_max_scaler.fit_transform(df_train[['LatitudeDeparture']].values)

df_train[['LatitudeArrival']]= min_max_scaler.fit_transform(df_train[['LatitudeArrival']].values)


df_train.head()

Date  Week  Weekend Celebs DateOfDeparture Departure  CityDeparture  \
0     1    49        0             2011-12-05       ORD        Chicago   
1     5     9        0             2013-03-01       ATL        Atlanta   
2     1     6        0             2012-02-06       ORD        Chicago   
3     1    50        0             2012-12-10       ORD        Chicago   
4     5    15        0             2012-04-13       SFO  San Francisco   

   LongitudeDeparture  LatitudeDeparture Arrival    CityArrival  \
0           41.978603           0.671019     EWR         Newark   
1           33.636719           0.738700     LGA       New York   
2           41.978603           0.671019     BOS         Boston   
3           41.978603           0.671019     SFO  San Francisco   
4           37.618972           0.000000     JFK       New York   

   LongitudeArrival  LatitudeArrival WeeksToDeparture    std_wtd  PAX  
0         40.692500         0.938417              8-9   5.667243    7  
1         40.777245         0.944181            10-11  10.001754    7  
2         42.364347         1.000000             9-10   7.136821    7  
3         37.618972         0.000000              8-9   7.404291    7  
4         40.639751         0.946004            14-15   9.858544    7

Δηλαδή έχετε διαθέσιμα τα δεδομένα απο 8899 πτήσεις για τις οποίες ξέρετε την κατηγορία του αριθμού επιβάτών `PAX`.

Αντίθετα, το αρχείο με όνομα `test.csv` περιέχει το test set του προβλήματος. Όπως και στην περίπτωση του training set, σας δίνεται και τώρα κώδικας ο οποίος φορτώνει το test set σε ένα DataFrame της βιβλιοθήκης Pandas και τυπώνει τις πρώτες 5 γραμμές του.

In [5]:
df_test = pd.read_csv('dataset/test.csv')
df_test.head()

DateOfDeparture Departure      CityDeparture  LongitudeDeparture  \
0      2012-10-21       DFW  Dallas-Fort Worth           32.896828   
1      2012-09-13       LAX        Los Angeles           33.942536   
2      2012-09-04       ORD            Chicago           41.978603   
3      2012-08-13       DEN             Denver           39.861656   
4      2012-09-10       ORD            Chicago           41.978603   

   LatitudeDeparture Arrival    CityArrival  LongitudeArrival  \
0         -97.037997     SFO  San Francisco         37.618972   
1        -118.408075     ATL        Atlanta         33.636719   
2         -87.904842     IAH        Houston         29.984433   
3        -104.673178     PHX        Phoenix         33.434278   
4         -87.904842     SEA        Seattle         47.449000   

   LatitudeArrival  WeeksToDeparture    std_wtd  
0      -122.374889         14.600000  11.575837  
1       -84.428067         14.730769  13.364304  
2       -95.341442          8.470588   5.885551  
3      -112.011583          8.200000   6.292853  
4      -122.309306         12.090909   9.138662

In [6]:
dates = df_test[['DateOfDeparture']]
print(dates.shape[0])

datetime=[p.parse(dates.iloc[i]['DateOfDeparture']) for i in range(dates.shape[0])]

date_number=[datetime[i].isoweekday() for i in range(dates.shape[0])]

date_month=[datetime[i].isocalendar()[1] for i in range(dates.shape[0])]


date_number=pd.DataFrame(date_number,columns=['Date'])
date_month=pd.DataFrame(date_month,columns=['Week'])

df_test['Date']=date_number.values
df_test['Week']=date_month.values


score = np.zeros((dates.shape[0]),dtype=int)
for i in range(dates.shape[0]):
    if datetime[i].isoweekday()==6 or datetime[i].isoweekday==7:
        score[i]=1
    
scor=pd.DataFrame(score,columns=['Weekend'])
df_test['Weekend']=scor.values


score = np.empty((dates.shape[0]),dtype=str)
for i in range(dates.shape[0]):
    if datetime[i].isoweekday() == 4 and 22 <= datetime[i].day < 29 and datetime[i].month==11:
        score[i]='Independance Day'
    if 24 <= datetime[i].day <= 25 and datetime[i].month==12:
        score[i]='Christmas'
    if (datetime[i].day == 31 and datetime[i].month==12) or (datetime[i].day == 1 and datetime[i].month==1):
        score[i]='New Year'
    
scor=pd.DataFrame(score,columns=['Celebs'])
df_test['Celebs']=scor.values

df_test[['WeeksToDeparture']]= pd.cut(df_test['WeeksToDeparture'],  [0,3.6,4.01,6,8,9,10,11,12,13,14,15,16,17,18.3,20.5,24],labels=[ '0-3', '3-4','4-6','6-8','8-9','9-10','10-11','11-12','12-13','13-14','14-15','15-16','16-17','17-18','18-20','20-24'])

cols = df_test.columns.tolist()
cols =cols[-4:]+cols[:-4]
df_test=df_test[cols]

df_test[['LatitudeDeparture']]= min_max_scaler.fit_transform(df_test[['LatitudeDeparture']].values)

df_test[['LatitudeArrival']]= min_max_scaler.fit_transform(df_test[['LatitudeArrival']].values)


df_test.head()

2229


Date  Week  Weekend Celebs DateOfDeparture Departure      CityDeparture  \
0     7    42        0             2012-10-21       DFW  Dallas-Fort Worth   
1     4    37        0             2012-09-13       LAX        Los Angeles   
2     2    36        0             2012-09-04       ORD            Chicago   
3     1    33        0             2012-08-13       DEN             Denver   
4     1    37        0             2012-09-10       ORD            Chicago   

   LongitudeDeparture  LatitudeDeparture Arrival    CityArrival  \
0           32.896828           0.493226     SFO  San Francisco   
1           33.942536           0.077221     ATL        Atlanta   
2           41.978603           0.671019     IAH        Houston   
3           39.861656           0.344594     PHX        Phoenix   
4           41.978603           0.671019     SEA        Seattle   

   LongitudeArrival  LatitudeArrival WeeksToDeparture    std_wtd  
0         37.618972         0.000000            14-15  11.575837  
1         33.636719         0.738700            14-15  13.364304  
2         29.984433         0.526253              8-9   5.885551  
3         33.434278         0.201740              8-9   6.292853  
4         47.449000         0.001277            12-13   9.138662

Όσον αφορά τις διαστάσεις του test set, όπως βλέπετε παρακάτω, αποτελείται από 2229 γραμμές και 11 στήλες. Η στήλη που λείπει αφορά τη μεταβλητή `PAX` την οποία στόχος σας είναι να προβλέψετε.

In [7]:
df_test.shape

(2229, 15)

### Στόχος

Όπως αναφέρθηκε και παραπάνω, στόχος σας είναι να προβλέψετε την κατηγορία αριθμού επιβατών (μεταβλητή `PAX`) για κάθε πτήση του test set. Πρόκειται για ένα supervised learning πρόβλημα. Θα πρέπει να επεξεργαστείτε τα δεδομένα του training set και στη συνέχεια, θα τα χρησιμοποιήσετε για να εκπαιδεύσετε κάποιο ταξινομητή, τον οποίο θα χρησιμοποιήσετε για να κάνετε προβλέψεις σχετικά με τα δεδομένα του test set.


### Αξιολόγηση

Η αξιολόγηση των προβλέψεων θα γίνει με βάση το `micro F1-score`. Το F1-score παίρνει τιμές μεταξύ 0 και 1. Όσο μεγαλύτερη η τιμή του F1-score, τόσο καλύτερος ο ταξινομητής που παρήγαγε τις προβλέψεις. Για κάθε μια από τις 8 κατηγορίες του test set μας, ο ταξινομητής θα κάνει κάποιες σωστές προβλέψεις και κάποιες λάθος. Πιο συγκεκριμένα, για μια κατηγορία $c$:
- τα παραδείγματα που ανήκουν στην κατηγορία $c$ και ο ταξινομητής προέβλεψε ότι πράγματι ανήκουν στην κατηγορία $c$ ονομάζονται True Positives (TP).
- τα παραδείγματα που ανήκουν στην κατηγορία $c$ και ο ταξινομητής προέβλεψε λανθασμένως ότι ανήκουν σε κάποια άλλη κατηγορία ονομάζονται False Positives (FP).
- τα παραδείγματα που ανήκουν σε κάποια κατηγορία διαφορετική της $c$ και ο ταξινομητής προέβλεψε ότι πράγματι ανήκουν σε κάποια κατηγορία διαφορετική της $c$ ονομάζονται True Negatives (TN).
- τα παραδείγματα που ανήκουν σε κατηγορία διαφορετική της $c$ και ο ταξινομητής προέβλεψε ότι ανήκουν στην κατηγορία $c$ ονομάζονται False Negatives (FN).

Δεδομένων των παραπάνω ορισμών, μπορούμε συνήθως να υπολογίσουμε μερικές μετρικές απόδοσης του ταξινομητή μας, λαμβάνοντας υπόψη τον συνολικό (για όλες τις κατηγορίες) αριθμό True Positives (TP), False Positives (FP) και False Negatives (FN).
Στο πρόβλημα που σας έχει ανατεθεί υπάρχουν συνολικά 8 κατηγορίες $c_1, c_2, \ldots, c_8$ οπότε μπορούμε να υπολογίσουμε:
Πρώτα, το `Recall` το οποίο ορίζεται ως:
$$ Recall = \frac{\sum_{i=1}^8 TP_{c_i}}{\sum_{i=1}^8TP_{c_i}+\sum_{i=1}^8FN_{c_i}} $$
και μετράει συνολικά το ποσοστό των παραδειγμάτων που προβλέφτηκαν ότι ανήκουν στην κατηγορία $c_i$ σε σχέση με αυτά που ανήκουν πράγματι στην κατηγορία $c_i$. Έπειτα, το `Precision` το οποίο μας λέει πόσα από τα παραδείγματα που προβλέφτηκαν ότι ανήκουν στην κατηγορία $c_i$, ανήκουν πράγματι σε αυτή την κατηγορία, και ορίζεται ως εξης:
$$ Precision = \frac{\sum_{i=1}^8 TP_{c_i}}{\sum_{i=1}^8TP_{c_i}+\sum_{i=1}^8FP_{c_i}} $$

Το `micro F1-score` ειναι ο αρμονικός μέσος του Precision και του Recall, και ορίζεται ως εξής:
$$ micro\text{ } F1-score =  \frac{2*Recall*Precision}{Recall + Precision} $$
Πρόκειται για μια ευρέως διαδεδομένη μετρική απόδοσης η οποία είναι και αξιόπιστη δεδομένου ότι λαμβάνει υπόψη τόσο τα False Positives όσο και τα False Negatives.

Αφότου υποβάλετε μια λύση στην πλατφόρμα (όπως περιγράφεται παρακάτω), το micro F1-score της λύσης σας υπολογίζεται αυτόματα και εμφανίζεται στην οθόνη.

### Υποβολή Λύσης

Αφότου έχετε σχεδιάσει και τρέξει τον αλγοριθμό σας, και έχετε προβλέψει την κατηγορία επιβατών για κάθε πτήση του test set, πρέπει να υποβάλετε την λύση σας στην πλατφόρμα Kaggle ώστε να αξιολογηθεί. Συγκεκριμένα, στην κεντρική οθόνη υπάρχει η επιλογή `Submit Predictions` η οποία σας δίνει τη δυνατότητα να ανεβάσετε τις προβλέψεις σας. Οι προβλέψεις σας θα πρέπει να συμπεριληφθούν σε ένα αρχείο όπου κάθε γραμμή περιέχει την πρόβλεψή σας για την πτήση η οποία βρίσκεται στην ίδια γραμμή του test set. Για παράδειγμα, παρακάτω σας δίνονται οι τρείς πρώτες πτήσεις του test set:
``` html
2012-10-21,DFW,Dallas-Fort Worth,32.896828,-97.037997,SFO,San Francisco,37.618972,-122.374889,14.6,11.5758369028
2012-09-13,LAX,Los Angeles,33.942536,-118.408075,ATL,Atlanta,33.636719,-84.428067,14.7307692308,13.3643037748
2012-09-04,ORD,Chicago,41.978603,-87.904842,IAH,Houston,29.984433,-95.341442,8.47058823529,5.88555060146
```
Η πρώτη γραμμή του αρχείου που θα υποβάλετε θα πρέπει να περιέχει την πρόβλεψή σας για την κατηγορία αριθμού επιβατών της πρώτης πτήσης, η δεύτερη γραμμή θα πρέπει να περιέχει την πρόβλεψή σας για την δεύτερη πτήση και αντίστοιχα η τρίτη γραμμή θα πρέπει να περιέχει την πρόβλεψή σας για την τρίτη πτήση.

**Προσοχή**: Στο αρχείο που θα υποβάλετε, κάθε γραμμή θα πρέπει να περιέχει 2 αριθμούς, έναν για το `Id` της πτήσης του test set και έναν για το `Label` που υποδηλώνει την κατηγορία επιβατών (αριθμοί 0-7). Επιπλέον, θα πρέπει να υπάρχουν συνολικά **2229** γραμμές όσες και οι πτήσεις για τις οποίες πρέπει να κάνετε πρόβλεψη. Τέλος, η κωδικοποίηση του αρχείου θα πρέπει να είναι **utf-8**. Φυσικά θα υπάρχει και μία ακόμη γραμμή στην αρχή που περιέχει τό όνομα των πεδίων `Id` και `Label`.

### Αρχικός Κώδικας Python

Σας δίνεται ένας αρχικός Python κώδικας ο οποίος χρησιμοποιεί ως μόνη πληροφορία το αεροδρόμιο από το οποίο αναχώρησε η πτήση και το αεροδρόμιο στο οποίο προσγειώθηκε για να προβλέψει την κατηγορία αριθμού επιβατών. Συγκεκριμένα, ο παρακάτω κώδικας αποθηκεύει στη μεταβλητή `y_train` τη στήλη `PAX` του training set, δηλαδή την στήλη την οποία θέλουμε να μάθουμε να προβλέπουμε. Στη συνέχεια, διαγράφει όλες τις στήλες του training set εκτός από τις `Departure` και `Arrival` οι οποίες αντιστοιχούν στους κωδικούς αεροδρομίων αναχώρησης και άφιξης. Επίσης για να είναι πιό εύκολος ο έλεγχος της λύσης σας (τοπικά), μπορείτε αν θέλετε να παράγετε ένα νέο training και test set απο το αρχικό training set, χρησιμοποιώντας την  `train_test_split` του `scikit-learn` (σχολιασμένη γραμμή παρακάτω).

In [8]:

# load data

from sklearn.preprocessing import LabelEncoder

X = df_train[['Week','Date','Weekend','Celebs','DateOfDeparture','Departure','CityDeparture','LongitudeDeparture','LatitudeDeparture','Arrival','CityArrival','LongitudeArrival','LatitudeArrival','WeeksToDeparture','std_wtd']]
Y = df_train[['PAX']]


le = LabelEncoder()
le.fit(X['WeeksToDeparture'])
X['WeeksToDeparture'] = le.transform(X['WeeksToDeparture'])
le.fit(X['DateOfDeparture'])
X['DateOfDeparture'] = le.transform(X['DateOfDeparture'])
le.fit(X['Celebs'])
X['Celebs'] = le.transform(X['Celebs'])
le.fit(X['Departure'])
X['Departure'] = le.transform(X['Departure'])
le.fit(X['CityDeparture'])
X['CityDeparture'] = le.transform(X['CityDeparture'])
le.fit(X['LongitudeDeparture'])
X['LongitudeDeparture'] = le.transform(X['LongitudeDeparture'])
le.fit(X['LatitudeDeparture'])
X['LatitudeDeparture'] = le.transform(X['LatitudeDeparture'])
le.fit(X['Arrival'])
X['Arrival'] = le.transform(X['Arrival'])
le.fit(X['CityArrival'])
X['CityArrival'] = le.transform(X['CityArrival'])
le.fit(X['LongitudeArrival'])
X['LongitudeArrival'] = le.transform(X['LongitudeArrival'])
le.fit(X['LatitudeArrival'])
X['LatitudeArrival'] = le.transform(X['LatitudeArrival'])
X.corr()

C:\Users\Lefteris\Anaconda3\lib\site-packages\ipykernel_launcher.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if sys.path[0] == '':
C:\Users\Lefteris\Anaconda3\lib\site-packages\ipykernel_launcher.py:14: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  
C:\Users\Lefteris\Anaconda3\lib\site-packages\ipykernel_launcher.py:16: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pan

Week      Date   Weekend    Celebs  DateOfDeparture  \
Week                1.000000  0.002901  0.003531  0.055426        -0.133900   
Date                0.002901  1.000000  0.415068  0.000442        -0.009952   
Weekend             0.003531  0.415068  1.000000  0.029874         0.005279   
Celebs              0.055426  0.000442  0.029874  1.000000         0.010886   
DateOfDeparture    -0.133900 -0.009952  0.005279  0.010886         1.000000   
Departure           0.004729  0.000550  0.003318 -0.014054        -0.008173   
CityDeparture       0.008686  0.000239  0.002829 -0.004745        -0.000654   
LongitudeDeparture -0.000565 -0.018403  0.001502  0.006970        -0.019295   
LatitudeDeparture  -0.004625 -0.005799  0.002001  0.006593        -0.000211   
Arrival            -0.007380 -0.010947 -0.001385  0.006694         0.011053   
CityArrival        -0.011322 -0.009701 -0.002661 -0.013059         0.009055   
LongitudeArrival   -0.005210  0.009406  0.012331  0.022379        -0.002794   
LatitudeArrival    -0.005448  0.025082  0.011434  0.011339        -0.001224   
WeeksToDeparture   -0.090599 -0.049651 -0.028205  0.005645         0.039095   
std_wtd             0.085705  0.051894  0.064781 -0.016370         0.125821   

                    Departure  CityDeparture  LongitudeDeparture  \
Week                 0.004729       0.008686           -0.000565   
Date                 0.000550       0.000239           -0.018403   
Weekend              0.003318       0.002829            0.001502   
Celebs              -0.014054      -0.004745            0.006970   
DateOfDeparture     -0.008173      -0.000654           -0.019295   
Departure            1.000000       0.719488            0.257366   
CityDeparture        0.719488       1.000000            0.104746   
LongitudeDeparture   0.257366       0.104746            1.000000   
LatitudeDeparture   -0.333768      -0.202584            0.289339   
Arrival             -0.124821      -0.090325            0.052148   
CityArrival         -0.096380      -0.136966           -0.008896   
LongitudeArrival     0.012543      -0.042263           -0.036443   
LatitudeArrival     -0.037847      -0.041017           -0.127279   
WeeksToDeparture    -0.070353      -0.050082            0.075330   
std_wtd              0.122581       0.132701           -0.142707   

                    LatitudeDeparture   Arrival  CityArrival  \
Week                        -0.004625 -0.007380    -0.011322   
Date                        -0.005799 -0.010947    -0.009701   
Weekend                      0.002001 -0.001385    -0.002661   
Celebs                       0.006593  0.006694    -0.013059   
DateOfDeparture             -0.000211  0.011053     0.009055   
Departure                   -0.333768 -0.124821    -0.096380   
CityDeparture               -0.202584 -0.090325    -0.136966   
LongitudeDeparture           0.289339  0.052148    -0.008896   
LatitudeDeparture            1.000000 -0.000298     0.002322   
Arrival                     -0.000298  1.000000     0.713527   
CityArrival                  0.002322  0.713527     1.000000   
LongitudeArrival            -0.121348  0.224927     0.044968   
LatitudeArrival              0.183342 -0.351012    -0.229197   
WeeksToDeparture             0.114671 -0.096136    -0.082103   
std_wtd                     -0.135811  0.118909     0.147669   

                    LongitudeArrival  LatitudeArrival  WeeksToDeparture  \
Week                       -0.005210        -0.005448         -0.090599   
Date                        0.009406         0.025082         -0.049651   
Weekend                     0.012331         0.011434         -0.028205   
Celebs                      0.022379         0.011339          0.005645   
DateOfDeparture            -0.002794        -0.001224          0.039095   
Departure                   0.012543        -0.037847         -0.070353   
CityDeparture              -0.042263        -0.041017         -0.050082   
LongitudeDeparture         -0.036443        -

In [9]:
import pandas
import numpy
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import chi2
# feature extraction
test = SelectKBest(score_func=chi2, k=4)
fit = test.fit(X, Y)
# summarize scores
numpy.set_printoptions(precision=3)
print(fit.scores_)
features = fit.transform(X)
# summarize selected features
print(features[0:5,:])


[   93.401   740.159  1842.208  1792.917 11517.645   499.986   344.961
   292.807   502.604   365.252   343.649   296.733   445.855   401.518
   115.002]
[[  1.   0.   0.  95.]
 [  5.   0.   0. 547.]
 [  1.   0.   0. 158.]
 [  1.   0.   0. 466.]
 [  5.   0.   0. 225.]]


In [10]:

from sklearn.ensemble import ExtraTreesClassifier

# feature extraction
model = ExtraTreesClassifier()
model.fit(X, Y)
print(model.feature_importances_)

[0.147 0.109 0.035 0.011 0.152 0.036 0.034 0.034 0.034 0.038 0.04  0.042
 0.038 0.107 0.143]


C:\Users\Lefteris\Anaconda3\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
C:\Users\Lefteris\Anaconda3\lib\site-packages\ipykernel_launcher.py:6: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  


In [11]:
from sklearn.feature_selection import RFE
from sklearn.linear_model import LogisticRegression
# load data

# feature extraction
model = LogisticRegression()
rfe = RFE(model, 4)
fit = rfe.fit(X, Y)
print(fit.n_features_) 
print(fit.support_)  
print(fit.ranking_)  

C:\Users\Lefteris\Anaconda3\lib\site-packages\sklearn\utils\validation.py:761: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\Lefteris\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\Lefteris\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:460: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silence this warning.
  "this warning.", FutureWarning)
C:\Users\Lefteris\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\Lefteris\Anaconda3\lib\site-packages\sklearn\linear_model\log

4
[False  True  True  True False False False False False False False False
 False False  True]
[11  1  1  1 12  2  3  9  7  5  4  8  6 10  1]


C:\Users\Lefteris\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\Lefteris\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:460: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silence this warning.
  "this warning.", FutureWarning)


In [12]:
y_train = df_train[['PAX']]

from sklearn.model_selection import train_test_split
df_train, df_test, y_train, y_test = train_test_split(df_train, y_train, test_size=0.2, random_state=42,stratify=y_train)

df_train.drop(df_train.columns[[4,6,7,8,10,11,12,13,14,15]], axis=1, inplace=True)

df_train.head()


Date  Week  Weekend Celebs Departure Arrival
8482     6    36        1              DFW     ATL
7687     1     4        0              LAX     ORD
5923     1    43        0              ORD     DEN
6661     5    46        0              CLT     ATL
6891     2    47        0              DFW     LAS

Για να μπορεί να δουλέψει ένα αλγόριθμος ταξινόμησης είναι απαραίτητο το training και το test set να έχουν ακριβώς τον ίδιο αριθμό στηλών (ίδιες μεταβλητές). Συνεπώς, πρέπει και στο test set να διαγράψουμε όλες τις στήλες εκτός από τις Departure και Arrival. Αυτό γίνεται με τον παρακάτω κώδικα.

In [13]:
df_test.drop(df_test.columns[[4,6,7,8,10,11,12,13,14,15]], axis=1, inplace=True)

df_test.head()

Date  Week  Weekend Celebs Departure Arrival
2231     6    14        1              BOS     PHL
3837     5     1        0              BOS     PHL
7812     4    40        0              ATL     CLT
6721     2    50        0              MCO     ATL
8        5    39        0              LAX     ORD

Ένας τρόπος αναπαράστασης κατηγορικών μεταβλητών είναι το λεγόμενο one-hot encoding όπου υπάρχει διαθέσιμη μια μεταβλητή για κάθε πιθανή τιμή του χαρακτηριστικού και ανάλογα με την τρέχουσα τιμή του, μια από αυτές της μεταβλητές είναι 1, ενώ όλες οι άλλες παραμένουν 0. Για παράδειγμα, αν είχαμε κάποια μεταβλητή Weekday η οποία περιέγραφε τη μέρα που έγινε μια πτήση, θα είχαμε 7 μεταβλητές (π.χ. 1000000 για Monday, 0100000 για Tuesday κτλ.). Σημειώστε ότι με την one-hot encoding αναπαράσταση ο αριθμός των χαρακτηριστικών που προκύπτει είναι ίσος με τον αριθμό των διαφορετικών τιμών που παίρνει η μεταβλητή. Παρακάτω εφαρμόζουμε one-hot encoding στις μεταβλητές `Departure` και `Arrival`.

In [14]:
import numpy as np

X_train = df_train
X_test = df_test
y_train = np.ravel(y_train)

from sklearn.preprocessing import OneHotEncoder

enc = OneHotEncoder(sparse=False,categories='auto')
enc.fit(df_train)  
X_train = enc.transform(df_train)
X_test = enc.transform(df_test)
X_train.shape

(7119, 105)

Βλέπουμε ότι ο αριθμός των στηλών αυξήθηκε από 2 σε 40. Αυτό συνέβη γιατί υπάρχουν 20 διαφορετικά αεροδρόμια και συνεπώς χρειαζόμαστε 20 μεταβλητές για να αναπαραστήσουμε κάθε μια από τις μεταβλητές `Departure` και `Arrival` χρησιμοποιώντας one-hot encoding.

Έπειτα, εκπαιδεύουμε ξανά έναν ταξινομητή logistic regression για να προβλέψουμε τις κατηγορίες αριθμού επιβατών των δεδομένων ελέγχου. Επιπλέον, αποθηκεύουμε τις προβλέψεις μας στο αρχείο `y_pred.csv` στο δίσκο.

In [15]:
from sklearn.neural_network import MLPClassifier
import timeit

#uncomment to reproduce results
#state1=('MT19937', [1546483054, 3151018956, 3632092744, 2091789978,  300308011, 3542745596, 1822245025, 1174413095, 2428608518, 3408020893, 186189152, 1784643307, 1123961950, 2949380232, 2269177984, 1284984153, 1836948424, 2037122910, 4280926093,  668832025, 3502219249, 1607507087, 2700591644,  259236845, 1926529433, 2992208145, 2459758489, 3311500098, 3576850881, 3553266599, 4112505810, 4041970324, 3565462707, 2504468881,  559419937, 3025784872, 2700843702, 1408916777, 2310613230,  459764035, 2285996695, 2065197554,  184145929, 3178106040, 1917882254, 2005215896,  562343819, 2887230726, 3216981700, 1047902799, 290441309, 1027527140, 3042431784,  164481479, 2006784697, 2672820943, 1001115673, 3117775638, 1570980638,  976097910, 3846729418,  207373962, 1743630000, 2786698516,  219211758, 4081835815, 3780463990,  122563692, 2123453152, 3839302666, 4145930451, 1640563287, 2129130038,   94568188, 3165091254, 872072271, 1730591351, 1296742107, 1917505104, 4028874564, 241727059, 3603831312, 3318314193, 2713549101, 1508112095, 3726586859, 2691605214, 3302096163, 1932459512, 2504225942, 2634822334,  807051399, 2451215519,  709863246, 3440885796, 1430564546, 1895059023, 1517069095, 1312142432, 1935610792, 1000813329, 2495758106, 2114895070, 3707558754, 1825601197, 1282727493, 2028606782,  577902918, 3748949770, 1301979130, 2709221749,  330134370,  366737434,  407162035, 2174062097, 260978930, 3862165998, 1958506486,  704928745, 3091657828, 1670578870, 1014301100,  331527510, 1010209129,  993497065, 1414428778, 2958280117,   31256754, 2782410938, 1370544409, 4003167994, 3743111616, 1630061939, 2766006399, 3831668439, 2847153707, 2001917877, 1057648269, 4129428267, 3184129875, 3846171777, 4149731799,  978488626,  612333385, 2887721309, 2394751500, 1829588504, 3552194928,  185667059, 1675005044, 1745029055,  109624461, 1689573689,  337685425, 3399251567, 2014217527, 2407429354, 1540114469, 3006249682,  410911919, 251822251, 3951380248, 2489286985, 3538691642, 2939065889, 3937335156,  834097561, 2133660420, 3977850529, 2740066451, 914377439, 1221058982, 1311729295, 3967251635, 2841852446, 1426254267, 4115710226, 1940211046, 1814659413, 3689020119, 4214188632, 3323767196,  929956721, 2186663244,  405137022, 1153392239, 2248876064,  260300837, 1862155349, 4007838177, 1046900968, 1731114055,  701535838, 3830817239, 3324988774, 987207324, 2230052944, 2728692511, 1126721591, 3224372245, 3921372022, 2472205042, 4271034746, 3361040008,  794744995, 584037404, 2685995994,   69223687, 3168598931, 3536247302, 2772677067,  374912544, 3310812020, 2082792904, 3456514595, 3027242871, 4037981953, 1307330211, 1987219652,  470780820, 2766161472, 3588426471, 1604538578, 1534381854, 1439059739, 424157219,  932997553,  891055288, 3278724476, 3282557696, 1182991381, 2070630603, 2367224474,   52311777, 3530625199, 2369495239,  586341285, 1591670278, 1842466993, 2437252703, 1843447969, 2708260881, 1669471089, 3809723427,  128024980, 1813795929, 1052270510, 1874058909,  979128452, 1360947469, 915070646, 3035841737,  425183251, 1983454076, 2963848783, 2961525344, 1511181226, 1310056056, 2751367103, 1547350932, 2455148493, 2383457456, 3386729792,  677302134,  514326678, 1286767159,  563496536, 3092699587, 2501063473,   22970156, 646625642, 1004200417, 3647178709,  114302591, 2045181997, 1215209071, 2205114490, 3101247597,   75889633, 1271843292, 1403513417, 2778199937, 2969486025,  575569458,  366906582, 3691386763,  364319174, 2384919101, 1560638203,  981739459, 700451857, 3216039128, 1223544102,  413978760, 1266618062, 2209783762, 3510853176,  345948528, 2721530458, 2192983523, 4067818481, 3355371943, 3860263650, 2670673140, 2281485374, 550189277, 2074522979, 1300562653, 2818956800, 1479387647, 133103724, 2608396243,  101687213,  442013050, 1531397212, 1094631148, 2255038909,  118337688,  269485366, 3583374733, 594862150,  856421087, 1086925373, 2929444591, 2909098437, 3526214600, 1751311197,  573829267, 3795949127,  690400797, 3996858299, 3560338659, 2418539948, 2655728883, 3516390755, 2214770735, 2396897297, 1305211856, 2913536967, 1767969548, 2279176821, 2291663176, 2108628360, 3886283108, 1778025979, 56600315, 4193662305,  778208005, 3055443797, 1784819120, 1862713651, 3412851737, 3896140194, 1957605094, 3235225989, 1516466497, 4187418532, 1079897416, 4060721203, 3194762327, 1392845809, 3167919897, 4093327889, 3803155589, 3240110154, 1295906618, 1230905013,  394031731, 2189276111,  661366354, 1303937484, 1876749396, 2935340029, 1082631184, 3032995115, 1678314404, 1011869329,  132157870, 2907805984, 2484248037, 657058975, 1870951224,  326760443, 2695965318,  256011156, 4057114341, 3149328448, 1603149197, 2679477696,  538372367, 2670459761, 3121594598, 1286414204, 2123518682, 3859735793, 386127621,  532185989, 4140896774, 3712207495, 3434843043, 1964423088,  203109222, 3305561808, 2226771410, 1768795300, 2653711278, 2933491826, 1625736647, 2873949878, 2962936477, 1579495765, 3579734975, 4161543880, 3570387636, 3900173009, 3887597689, 2608505794, 1933859937,  369611650, 2178221549, 918634479, 2566791664, 1644947488, 1389576108, 1948693993, 1677127217, 1496558234, 3857466322,  758684705, 1405764530, 2867626573, 3845435866,  166088013, 3226842642, 3870748519, 2818897959, 1472509261, 3937744049,  462360048,  259387751, 1166630491, 2133222715, 4075289500, 4035403382, 1210573541, 879004081, 3912306067, 1603038095, 2966908614, 3332479253, 1346907504,  920807512, 3415360636,  181796576, 1140027430, 2602639402, 2942245264, 1613904483, 3317529716,  111070142, 1373704642, 2158124732, 1677004996,  655101064, 2452470648, 3182772979, 1508729831, 1214808465, 3380686756, 1678323128, 2904952019,  848685388, 2685581524, 2183623751, 3232550419, 3773029931, 2665615268,  521798747, 3021302981, 1897622370, 4294859503,  712387837, 4120782515, 3639178579, 2031295092, 317001486, 1523544172, 3858543080, 4183160479, 3995584885, 2327618424, 2688469773, 3034195415, 4246924278, 1589655191, 86364445,  949946209, 2438452278,  362974838,  882465921, 3330902489, 2885619447,  525665695,  400969906, 4226488882, 2552008782, 1318165494, 1672132974, 2510328263,  247394742, 1688847564, 1860656096,   10070725, 1080479418, 4165903561, 3334283189, 1150282450, 2167902788, 1075719588, 1306022176, 2645142797,  284643316,  721953102,  623502289, 2066267265, 2339561613, 4134019449, 1398112561, 1798220288, 1416273270, 3734360325, 1025617777, 2638525097, 3596944012,  169970561, 795546876, 3864929668, 1072369500, 1048137062, 3054575961, 1293475331, 2746002151, 1121385335,  329080237, 2146602081, 2658190337, 2071458385, 1194290099, 2151006558, 1932756849, 4208943446,  176874928,   92786840, 2975770145, 4030023929, 3219756749,  746997975,  534171136,  637356078, 3922729557, 2643225630,  319137085, 3344095299, 3811152755,  660479332, 3818397353, 3256030536,   16354254, 2833737854, 2778371877, 2800911261, 1400309494,   91287727, 4280655432, 3281688789, 1570952365, 1975645724,  956735666, 2289740412, 1326756857, 3544365852, 1095789696,  844290603, 1285429310,  319965731, 450428184, 3266688194, 1823644528, 1673383649,  364413101, 3479329679,   94375051, 3668465703, 4103212165, 2247928112, 3415770381, 2109931226, 1414447548, 1894487783, 3372065045, 3068089862, 3040941549, 3304296613,  151710169, 1499620857, 1445721397,  704418786,  519589001, 4239033709, 4272981191, 1141721014,  513752214, 1012477842, 1856170239, 1541921948, 2614432088, 4073653482,   61645654, 2514450776,   43125997, 2198139117, 3228301240, 1548276277,  279233779], 624, 0, 0.0)
#state2=('MT19937', [2872343671, 2239023895, 1989005992, 2519389960, 1606100909, 3718771305, 1842980521, 2910468440, 3807414888, 1791789786, 796701347, 1373523649, 2316986294, 1648370832,   80477298, 2429270491, 1461550353, 3135038352, 1057777923, 3282306136, 2066801348, 3507729908, 3363055919,  235873889, 3687290411, 1629795779, 1449038583, 2347310395, 3301081224, 2739876235, 390198169, 2024285591,  772626887, 3462221287, 2717173637, 1548630915, 3095782304, 3463095357, 1575543138, 3672058273, 124560235,  248673361, 2663051704, 1518250015, 1307946063, 4108870795, 2132301863, 2596842470, 3717329718, 3142873513, 3826945492, 2480090006, 4161855916,  825588097, 3627521505, 4185330975, 2714407537,  532143429, 3805926185, 1752997498, 1943772472, 1520634692,  304769174, 2787480520, 1475739457, 3042506024, 1682214385,  293467540, 1860510046, 1505627001, 389311173, 1288126802,  915714518,  470975160, 1388767631, 193983247, 1912840774,  774566555, 2861275107, 2567286045, 3555208892, 2117452167, 1163247094, 1934251635, 1776123907, 2054280122, 1662002180, 1551175746, 1154510021, 1370067042, 857649990,  849707466, 1268862530, 1134419405, 2338487878, 1708639166, 1275266410, 3555368935, 3172739332, 3025233296, 112495285, 1483577202, 1032804170, 3263930936, 3465733140, 2528042874,  719213582, 3752763154,  504200721, 1038809666, 2046047289, 3743054485, 2452866448, 2480590673, 2034047929, 1909631940,  973321500, 2444465768, 3137021070,   67128704, 3220329826, 1930231524, 3613011773,  481016976, 1546476627, 1519536415, 2252125896, 1466506080, 2017905669, 2095481875, 3114227650, 1672869881,   78803962, 4224855099, 4046761469, 1486401653,  406519656, 3008273607, 3810100136, 3661869680, 1274863943, 3433192237, 2336228046,  382682823, 1176801592, 1085641218, 2916957023,  119247612, 1202067749, 2576529665, 2751064786, 2683068573, 3682017146, 1304037380, 1681904478, 1345271897,    6508825, 3840877025, 1328677034,  601718995, 1859151099, 4033070288, 3485059421, 1915480481,  975424384, 3440266098, 3600691637, 1524804385, 3802778272,   39272285, 2773916078, 1750671338,  716893672, 2031871706, 1195031601, 4110632070, 3882134969, 1402460760, 3207733890, 1883333755, 4137608199,  348038680, 2220999327, 1800913868,  348953075, 2364936164,  610485089, 1128920976, 1664189106, 2358652414, 2563886991,  343834205,  545281667,   52686237, 3068336899, 4081237891, 4104395190, 1670992989, 2982790392,  488323580, 2785494253,  190176421, 4048793508, 2060347298,  245680814, 1866605039,  346514631, 3423184827, 1504735352, 3286259045, 1031314860, 1942184226, 2383521544, 3483813359,  902055134, 135453513, 2686786406, 2147280730, 2803123838, 4050720510, 553238214, 1875361570,  566524292, 1026041051, 3516769215, 4032180353, 2672550932, 4059192061, 3045006501, 1244605551, 2036326790, 4082944913,  785641338, 4083686186, 2322571560, 1109718107, 3571190294, 1193726502, 2869637704,   17489544, 4249129456,  456865682, 4148921342,  637503672, 3119095100, 4002106526,  502618758, 2187733461, 4270217904, 1289332633, 1614369510, 1154292621,  216096049, 1957551274, 4187998553, 3851893137, 1174560490, 1060156418, 2960365849, 3890590327, 2673761163, 3646023922,  866899706, 2449946864, 3733760935, 1410500635,  935319656, 1172451950, 4280919166, 1662547212, 2110056853, 3413047884, 2478922723, 4163224373, 2245361902, 4201517562, 2601408883, 3396927748,  829088578, 2487723015, 3371613627, 1805067541, 1400171988, 1601857192, 2489530122, 668439939, 1884020407, 2597456880,  203260025,  421749976, 2587425822, 2759751809, 2644214464, 1237016592, 3724162116, 960373592, 2943380332, 3589358719,  522990276, 1097821437, 4176601694, 4112626246, 1175283715, 3124759887,  561645100, 2172510266, 3981622224, 3571048916,  773731522, 2708598450, 4067525585,  184306945, 1428619104,  132051894, 1535637517, 2233696468,  418282044,  331155217, 3794275742, 1198460481, 3823212525, 2275235606, 2291344879, 2193899801, 3027829254, 54382899,  865923616, 2559568275, 3644805939, 2256882089, 1567994619, 4182970977, 1143315281, 3701379937, 1119619279, 816934812, 3303335605, 3337461044, 2946511280, 4292967332, 2976154358, 1217864109,    5067080,  643468278, 1968792208, 678416194,  453333279, 3419363461,  311338391, 1214503335, 2685465455, 1362913456, 3573002820, 2999336700, 1747213598, 1128953835, 3396784264, 2373083026,  221021612, 2911333571, 3218348416, 1737750997,  279975620, 1579961456,  244737626, 3051554159, 3062737950,  201231784,  692399554,  103397960, 505858314, 4203117537, 1564039135, 2046988667, 3820087752, 3580101148, 2867799064, 1560243920, 2225926282, 2077020627, 3991375190, 3664005258, 1051607519, 3302282865, 3514916560, 401629493, 3716444522, 2361213363, 1737906724, 3281826121, 4239300873,  969394051, 3559104718, 2746224447, 3936735189, 3540916985, 1162525714, 1680880092, 1592463450, 4038655080, 1218628598, 3960865899, 2994607193,  626996189, 2415614079, 2573234487, 1824120655, 1560701690, 3219150326, 1260742857, 219924826, 1983641861, 4138402062, 1433495724, 3378480368, 2935018391, 1991818446, 3527378612, 2399267926, 3235975397, 1937894858, 3964934003,  876826023, 2070613882, 2911891147, 2877770656,  435301236,  874313811,  841384658, 1190793495, 625235540, 4127700326, 2742804206, 2085658230, 3767735994, 165740752, 1263627920, 1369476132, 3890210925,  536847811, 490731576, 1806130311, 2182337857, 1621619574, 3750415221, 2937236915, 1840007646, 1756264508, 3537544944, 1275342429, 600822476, 4093014837, 3040058922, 3549713437, 3094487732, 3298361209, 3767875011, 3614070489,  273457233, 1269848997, 1539900681, 1236172784, 1016574714,  785177614, 2889308382, 239925473,  628350545,  322601570,  528235190,  791570180, 819553558, 2761521320, 1494699549, 2971014835, 3598260114, 3447309517, 3801870998, 1315674109, 3452879500,  125323729, 150685509, 1945057444, 1602952519, 1171706958, 3714989465, 2451375305, 2226082128,  167559914, 3638164108, 1846699012, 2847590237,  492443321,  518546710, 1530666363, 1855059200, 554817445, 2098638243, 3048852471, 1623464716,  824667429, 2410210184, 3029332487,  179320634,  972736749, 3067177141, 4245991787, 3126114264,  726190481, 3345048313,  845358113, 1573285908, 1522157609, 1056559185, 1603990601, 4003136040, 2327378734, 3147652809,  609278675, 4150308359,  432741536, 3699214162,  726269077, 3579135054, 1094004351, 1579077086, 3647958466, 2541259356, 1726695449, 1218609475, 2378549859, 1189672992, 1131351963, 2674455801, 2259579303, 1648379464, 1178154438, 1084995162, 1837440578, 2668764331, 2727473305, 3384791022, 3874358880, 2775894523, 1045910750, 2161478179, 1971904187,  306066589, 1715485018, 1076327451, 1941842975, 147249575, 1174333010, 1035303606, 2266202353, 3554713895, 3284538924, 1831080762,  217115355, 3223460337, 3483943785, 2980907826, 3740329512,  990958984, 1942751237, 4028581680, 2670153083, 3460373537,  328215257, 3297560526, 1896870723, 3731249177, 3975637406, 3400466463, 3211314440, 3136634400, 526953092, 1658964932, 1938691023, 2458654626, 3650816692, 3061976247, 2784498231,  992979644, 1078398631, 3800180309, 2568765309,  832386294, 2273144750, 3091269997,  907870210, 657672074,   45109165,  900580771, 2972537174, 1098311713, 3821436755, 2460509233, 1783159618, 2556908948, 1644017023, 429207801, 3182413837, 2442009295,  589954039,  368380036, 2825496450,  775673615, 1131700493, 3483610443, 2623846017, 19786647, 4153878607, 1962373074, 3952548886,  434727237, 1761500584, 3383161218, 3330604458, 1211598432, 1514838979, 35406292,   75391790, 1890474237, 2312647414], 44, 0, 0.0)
#state3=('MT19937', [2108211128, 2527430299, 4192017223, 2625402879, 3174555681, 2022678817, 2680360192,  687524803,  789983273, 2559032048, 1061351843, 3907993216, 2274468456, 1646363939, 2553621988, 1967080462, 3390661242,  489435987, 3377982090, 3819426765, 2403597218, 2400012891, 1091163126, 3797853834, 4185460117, 4031381599, 1162567273, 1491538376, 3389268462, 3579154880, 2813542789, 2779274534,  214652136,  974169338, 3783704622, 2615002183, 3808356746, 1186467645,  654087330, 2572617128, 292098669, 1623059241, 3516970443, 3046522805, 2182150295, 2408924925,  727368013, 3280097587,  806363758, 2238248469, 1981245307,  964950666, 3982297489, 3290915655, 1890884893, 181690625, 2599059891, 1179164902, 2553705857, 1059991057, 4199608049, 2001070458, 3294358102,  306333077,  723826508, 1611778045,  399012637, 3951775085, 2321903680, 1777277873, 98125490,  511201609, 1814560373, 2576718381, 4158404049, 1433640578,  920052697, 3251859353, 4133599423,  782901469, 2993506619, 1071111702, 1925203904, 1660956907, 1686103911, 494434633, 3596361235, 1574842614,  945643900, 2191525063, 105156582, 3652152993,  968037704, 3137791607,  865559392, 3945040168, 1141623331, 1430622273,  110623811, 1718518439, 1035216736, 1087457138, 3868303794, 2909754173, 1972200715, 3681844293, 3608669568, 2423844206, 1504309087,  657859171, 3488251429, 3014939202, 4132001620, 3850607644, 1139049802, 1082107566, 1275988679, 1086467220,  804164628, 2281639504, 3615059499, 4160469798, 2736290919, 3363335257, 3417667264, 4061779585, 3395559934, 2146057535, 1974337447, 2999032146, 2312036576, 4166427877, 2510950347, 2588072156, 3391743315, 396258423, 2099316707, 3828309440, 4065148968, 1400340539, 3311742139, 1772401249,  183195249, 4131114934, 2571570611, 2639323152,  689610514, 3347120954, 3318589318, 3533537723, 178825461, 3149458077, 1738672616, 1883281519,  486825440, 3042045025, 1192697823, 2516415278,  999934003, 4218100339, 3961921282, 3478708235,  418698777, 2027724052, 3457585540, 1952805688,  725580342, 2067074748,  986043452, 3004018845, 1581084803, 1783939526, 4120508539,  379304578, 2476788011, 1904556551, 2837766117, 1457212402,  595024117,  601163942, 4055770621, 2628986244, 3461380316,  924710345, 1046943985, 1232467372, 2548339570, 3442033102, 3257747650,  618783007, 1810556078, 2757258952,   92263296,  977718763, 4045339016, 4050996239, 4082130739, 1024033951,  403275062, 2451238233, 2931697124,  190791769, 1290298498, 1842249583, 2886765473, 3808813389, 3633569851, 2723940308,  911780196, 2570721751, 795298783,  988269220, 1377512993, 3997777568, 3356629907, 4217693401, 2515842228,  553972077, 2192487042, 3995250221, 3797090983, 2745687960, 2764238015, 1939505332, 3392432398, 323737314, 3608612438, 2265358502, 2774960106, 1629014827, 61753951, 2469040051,  546757442,  585785223, 1823750681, 2716207879, 3261807884, 3849154842, 3067029964, 1658529734, 1803333941, 4077881701, 3559487879, 3002827510, 3699472086, 1314756722,  825508683, 1901493230, 4232452041,  680094928, 3067281358, 3322871163, 4088062714,  511036671, 1209002524, 4106518276, 1592519891, 3288681016,  181936271,  744489648, 557326180, 2784656729, 2612192269,  867343332, 2431853971, 2096331853, 3571112273, 1863569619,  746559439, 3433603465, 1406906813, 1813755522,  668223369, 3111633112, 2679142124, 2925976690,  484084533, 2801128897, 2075385091, 4049679332, 685192293,  924238170, 1316150886, 3129650176, 1710284877, 2149708551, 1365528054, 2654747247, 2548313685, 3879181058, 2906579212, 3499810762, 3305248716,  400089269, 3694511090, 2956212538, 3319643443, 1301117711, 4039294822,  391713788, 2427432361,  319636754, 3091952140, 1317429426,  905140594, 2703007861, 3176050470,   35626495, 3522457115, 2659274794, 4051746132,  659003405, 3945646157, 3350807117, 3066037377, 2048290665, 2499527840,  735403925,   97765384, 4158619651, 1084260382, 1857774622, 2597984738, 1471162735, 3240155620, 1769378467, 1172117815, 1681267160, 2204169107,  447996314, 3145107969, 2041966415,  557441701, 2451346704,  440982736, 2847720999,  256391291,  176593614, 2455716607,   48703866, 2378444445,  236555167, 3564783692,  785715672,  113081979, 2080828578, 1987503170, 2577402027,  977034130, 2453773068, 4289885312,  165612196, 3816125300, 1209091645, 3602196786, 2740753984,   62624679, 3773673455, 2726329129, 3504645825, 2216598383, 2704259081, 3744605582, 2158044142, 2701544741, 1607860008, 2308682230, 3956800591, 1547041200, 2820919929, 1339793351, 2265304255, 3679402875,  527186701, 2314833627, 3011445892, 3460879416, 1527791943,  881945393,  491681403, 1140952913, 1435974984,  573044504,  868367355, 4023300740, 316652323, 1346655812, 1369258211, 3340283052, 4023420326, 3245754279,  921853931, 4089688143, 1528643438, 2630389314, 3062538964,  460504303, 1695817083, 1085454948, 3349793710, 3828186247, 2520216885, 4208136182,  339431303, 1443438967, 2018226642, 1270306008,  771039641,  332479130, 4110799016, 455679854, 3323345398, 3431819626,   50095539,  687997097, 890761220, 1209528505, 3748522798, 2565241538, 2697849588, 1072583840, 3797698459, 2955238572,  784497924, 3735794565, 27428451, 2713189150, 4019579977,  417738450, 3084334794, 1194996394, 3652279977, 2507080558,  101481278, 2375028714, 2955067690, 1711609550, 4272039316, 3832490641, 3354958512, 3458372628, 3103983613, 1223856194, 1680998300, 2247996131, 1970828101, 3911091681, 1816830356, 3761314569,  238212915, 2513079548, 4047215625, 3516804536, 3183502235,  384782634, 174531601, 1673486073, 1576853331,  182031195,  395023627, 3730780759,  172328330, 2490846282, 2327297234, 1818232020, 4269248683, 4248977035, 3569924383,  909240122, 2605273357, 1998608194, 1510641474, 1641790130, 1172833791, 2522032074, 2171357350, 1717167697, 1190619983,  107030942, 3096952759, 3108852545, 1215931034, 4196006042, 3068024557,  261372072, 467529116, 2061592456,   63332911,  782276471,  826944233, 1821021470, 3563239388, 2533712700, 1905686700, 1458348968, 492358119, 2536567081,  544777088, 2340468877, 1793868296, 2818596855, 1526639114, 1308284632, 4016518676, 3639315966, 1400769388,   20271569, 2559166868, 3190616761, 4068831899, 641939100,  728692401, 3825929579,  421841888, 1707495467, 3871324694, 3324929206, 1643446389,  389579305, 3697673575, 1621372114,  429142037, 3212140329,   69212767, 1207744546, 1778464804,  595001801, 2740933585, 2257421564, 2239849438, 1940338110, 3710025581, 1988303760, 4064464883,  136684297, 337416451, 2398508180, 3383131018, 3288574218, 3747438177, 2238360023,  474174658,  750849410,  554690679, 2832180947, 3016105702, 1646209648, 3091808314, 1236224803,  434619131, 177209167,  624010569, 4025481646, 2799578942, 2609193957, 1658816594, 2726871629,  945244465, 1091392649, 3023235044, 3861410548, 3709494645,  656796139, 3430081417, 3344912019, 989366657, 1930649093,  783375650, 1454380084, 3129971147, 1033301807, 1247869205,  352989108,  791305219, 1078403938, 2458717678, 1545136945, 2940146123, 1430547443, 2701507027, 3885238974,  455411733,  536077325, 4007875855,  772681747, 3230830266,  187871694, 2607126894, 2876437473, 2032673873, 1741428585,  205756050, 3675082053,  471009104, 3451833413, 4163447888, 2673521296, 3284603168, 1259649339, 2487790146, 468866744, 1956220276, 2070821947, 1850974543, 3335171295, 3676813206, 2687343829,  241570264,  848224894,  503016090, 644703484, 3201812132, 2225511387, 1096607806, 2265814463, 2533212665,  271020420, 3381550487,  356587068, 3115740140, 3660253096, 2416512950,  305366375, 1705435955], 91, 0, 0.0)

total=0

#uncomment to reproduce results
#np.random.set_state(state1)

# Initializing the multilayer perceptron
mlp = MLPClassifier(hidden_layer_sizes=(15,5),solver='sgd',learning_rate_init= 0.008, max_iter=1000,tol=0.000000001,verbose=True)
#print(np.random.get_state())
start = timeit.default_timer()
mlp.fit(X_train, y_train)
stop = timeit.default_timer()
total=total+(stop-start)
y_pred1 = mlp.predict_proba(X_test)
pred1=np.argmax(y_pred1,axis=1)

#uncomment to reproduce results
#np.random.set_state(state2)

# Initializing the multilayer perceptron
mlp = MLPClassifier(hidden_layer_sizes=(15,5),solver='sgd',learning_rate_init= 0.008, max_iter=1000,tol=0.000000001,verbose=True)
#print(np.random.get_state())
start = timeit.default_timer()
mlp.fit(X_train, y_train)
stop = timeit.default_timer()
total=total+(stop-start)
y_pred2 = mlp.predict_proba(X_test)
pred2=np.argmax(y_pred2,axis=1)

#uncomment to reproduce results
#np.random.set_state(state3)

# Initializing the multilayer perceptron
mlp = MLPClassifier(hidden_layer_sizes=(15,5),solver='sgd',learning_rate_init= 0.008, max_iter=1000,tol=0.000000001,verbose=True)
#print(np.random.get_state())
start = timeit.default_timer()
mlp.fit(X_train, y_train)
stop = timeit.default_timer()
total=total+(stop-start)
y_pred3 = mlp.predict_proba(X_test)
pred3=np.argmax(y_pred3,axis=1)

y_pred = np.zeros((len(pred1)),dtype=int)


for i in range(len(pred1)):
    if pred1[i]==pred2[i] or pred1[i]==pred3[i] or pred2[i]==pred3[i] :
        if pred1[i]==pred2[i] or pred1[i]==pred3[i]:
            y_pred[i]=pred1[i]
        else:
            y_pred[i]=pred2[i]
    else:
        
        max1=np.amax(y_pred1[i])
        max2=np.amax(y_pred2[i])
        max3=-np.amax(y_pred3[i])
           
        if max1>max2 and max1>max3:
            y_pred[i]=np.argmax(y_pred1[i])
        elif max2>max1 and max2>max3:
            y_pred[i]=np.argmax(y_pred2[i])
        else:
            y_pred[i]=np.argmax(y_pred3[i])

print('Time: ', total)

Iteration 1, loss = 2.12736485
Iteration 2, loss = 1.92531943
Iteration 3, loss = 1.86394095
Iteration 4, loss = 1.83653446
Iteration 5, loss = 1.81918955
Iteration 6, loss = 1.80601367
Iteration 7, loss = 1.79487896
Iteration 8, loss = 1.78468262
Iteration 9, loss = 1.77487099
Iteration 10, loss = 1.76517459
Iteration 11, loss = 1.75507490
Iteration 12, loss = 1.74489206
Iteration 13, loss = 1.73408774
Iteration 14, loss = 1.72262050
Iteration 15, loss = 1.71087017
Iteration 16, loss = 1.69643892
Iteration 17, loss = 1.68098010
Iteration 18, loss = 1.66401023
Iteration 19, loss = 1.64479198
Iteration 20, loss = 1.62293531
Iteration 21, loss = 1.60008155
Iteration 22, loss = 1.57784608
Iteration 23, loss = 1.55322292
Iteration 24, loss = 1.53151351
Iteration 25, loss = 1.50931799
Iteration 26, loss = 1.48827240
Iteration 27, loss = 1.47040850
Iteration 28, loss = 1.45207993
Iteration 29, loss = 1.43772787
Iteration 30, loss = 1.42266740
Iteration 31, loss = 1.41009328
Iteration 32, los

Iteration 252, loss = 0.87837237
Iteration 253, loss = 0.87833249
Iteration 254, loss = 0.87763374
Iteration 255, loss = 0.87477984
Iteration 256, loss = 0.87501736
Iteration 257, loss = 0.87630151
Iteration 258, loss = 0.87389412
Iteration 259, loss = 0.87499262
Iteration 260, loss = 0.87250307
Iteration 261, loss = 0.87511369
Iteration 262, loss = 0.87084267
Iteration 263, loss = 0.87017150
Iteration 264, loss = 0.87171089
Iteration 265, loss = 0.87082058
Iteration 266, loss = 0.86986883
Iteration 267, loss = 0.87219652
Iteration 268, loss = 0.87253369
Iteration 269, loss = 0.87000477
Iteration 270, loss = 0.86901982
Iteration 271, loss = 0.87036109
Iteration 272, loss = 0.86688526
Iteration 273, loss = 0.86710417
Iteration 274, loss = 0.86766066
Iteration 275, loss = 0.86597982
Iteration 276, loss = 0.86403692
Iteration 277, loss = 0.86642167
Iteration 278, loss = 0.86482217
Iteration 279, loss = 0.86445144
Iteration 280, loss = 0.86276691
Iteration 281, loss = 0.86230807
Iteration 

Iteration 76, loss = 1.21971634
Iteration 77, loss = 1.21523378
Iteration 78, loss = 1.20952471
Iteration 79, loss = 1.20398409
Iteration 80, loss = 1.19877225
Iteration 81, loss = 1.19671849
Iteration 82, loss = 1.18980112
Iteration 83, loss = 1.18416037
Iteration 84, loss = 1.17859862
Iteration 85, loss = 1.17428584
Iteration 86, loss = 1.16865290
Iteration 87, loss = 1.16484689
Iteration 88, loss = 1.15924627
Iteration 89, loss = 1.15419184
Iteration 90, loss = 1.14935379
Iteration 91, loss = 1.14427958
Iteration 92, loss = 1.14000854
Iteration 93, loss = 1.13575923
Iteration 94, loss = 1.13003832
Iteration 95, loss = 1.12569446
Iteration 96, loss = 1.12245697
Iteration 97, loss = 1.11716297
Iteration 98, loss = 1.11295459
Iteration 99, loss = 1.10862664
Iteration 100, loss = 1.10489996
Iteration 101, loss = 1.10063275
Iteration 102, loss = 1.09709178
Iteration 103, loss = 1.09428664
Iteration 104, loss = 1.08844445
Iteration 105, loss = 1.08735839
Iteration 106, loss = 1.08269172
I

Iteration 331, loss = 0.85869382
Iteration 332, loss = 0.85847511
Iteration 333, loss = 0.86028959
Iteration 334, loss = 0.86074387
Iteration 335, loss = 0.85987204
Iteration 336, loss = 0.85830332
Iteration 337, loss = 0.85738998
Iteration 338, loss = 0.85699043
Iteration 339, loss = 0.85754420
Iteration 340, loss = 0.85889078
Iteration 341, loss = 0.85823005
Iteration 342, loss = 0.85520754
Iteration 343, loss = 0.85782951
Iteration 344, loss = 0.85660000
Iteration 345, loss = 0.85391658
Iteration 346, loss = 0.85253747
Iteration 347, loss = 0.85566488
Iteration 348, loss = 0.85499739
Iteration 349, loss = 0.85285928
Iteration 350, loss = 0.85199536
Iteration 351, loss = 0.85312139
Iteration 352, loss = 0.85365295
Iteration 353, loss = 0.85043505
Iteration 354, loss = 0.85042199
Iteration 355, loss = 0.85129732
Iteration 356, loss = 0.85247599
Iteration 357, loss = 0.85283604
Iteration 358, loss = 0.85085491
Iteration 359, loss = 0.85040658
Iteration 360, loss = 0.84721410
Iteration 

Iteration 189, loss = 0.90110019
Iteration 190, loss = 0.89763016
Iteration 191, loss = 0.89847359
Iteration 192, loss = 0.89667031
Iteration 193, loss = 0.89639470
Iteration 194, loss = 0.89471165
Iteration 195, loss = 0.89296885
Iteration 196, loss = 0.89569274
Iteration 197, loss = 0.89052292
Iteration 198, loss = 0.89066188
Iteration 199, loss = 0.88927442
Iteration 200, loss = 0.88832950
Iteration 201, loss = 0.88683464
Iteration 202, loss = 0.88740026
Iteration 203, loss = 0.88537319
Iteration 204, loss = 0.88407176
Iteration 205, loss = 0.88194993
Iteration 206, loss = 0.88171627
Iteration 207, loss = 0.88176469
Iteration 208, loss = 0.88071659
Iteration 209, loss = 0.87876899
Iteration 210, loss = 0.87671243
Iteration 211, loss = 0.87670176
Iteration 212, loss = 0.87720670
Iteration 213, loss = 0.87473390
Iteration 214, loss = 0.87393057
Iteration 215, loss = 0.87426151
Iteration 216, loss = 0.87199375
Iteration 217, loss = 0.87187222
Iteration 218, loss = 0.87311312
Iteration 

Iteration 440, loss = 0.78268265
Iteration 441, loss = 0.78096959
Iteration 442, loss = 0.78260738
Iteration 443, loss = 0.78024107
Iteration 444, loss = 0.78102701
Iteration 445, loss = 0.78351877
Iteration 446, loss = 0.77913994
Iteration 447, loss = 0.78052912
Iteration 448, loss = 0.77533743
Iteration 449, loss = 0.77877651
Iteration 450, loss = 0.77790068
Iteration 451, loss = 0.77922285
Iteration 452, loss = 0.77645403
Iteration 453, loss = 0.77893907
Iteration 454, loss = 0.77866510
Iteration 455, loss = 0.77812485
Iteration 456, loss = 0.77767440
Iteration 457, loss = 0.77772924
Iteration 458, loss = 0.77778552
Iteration 459, loss = 0.77867707
Training loss did not improve more than tol=0.000000 for 10 consecutive epochs. Stopping.
Time:  33.9093366


In [16]:
import csv

with open('y_pred.csv', 'w', newline='') as csvfile:
    writer = csv.writer(csvfile, delimiter=',')
    writer.writerow(['Id', 'Label'])
    for i in range(y_pred.shape[0]):
        writer.writerow([i, y_pred[i]])

In [17]:
from sklearn.metrics import f1_score

f1_score(y_test, y_pred, average='micro')

0.549438202247191

Υποβάλλουμε το αρχείο `y_pred.csv` στην πλατφόρμα και μας δίνει macro F1-score ίσο με 0.33. Στα πλαίσια της παρούσας εργασίας καλείστε να τροποποιήσετε τον παραπάνω κώδικα ώστε να προβλέψετε τις κατηγορίες αριθμού επιβατών των πτήσεων του test set. Μπορείτε να εφαρμόσετε κάποια μέθοδο επιλογής χαρακτηριστικών στα δεδομένα ώστε να κρατήσετε
μόνο ένα υποσύνολο από τα χαρακτηριστικά. Μπορείτε επίσης να δημιουργήσετε νέα χαρακτηριστικά τα οποία
ίσως βοηθήσουν στην κατηγοριοποίηση. Μπορείτε επιπλέον να πειραματιστείτε με κάποια μέθοδο μείωσης
διάστασης και να διερευνήσετε αν η εφαρμογή της βελτιώνει το αποτέλεσμα της κατηγοριοποίησης. Επίσης,
μπορείτε να χρησιμοποιήσετε θορυβώδη ή ανούσια χαρακτηριστικά για να παράγετε νέα χαρακτηριστικά που
παρέχουν μεγαλύτερα ποσοστά πληροφορίας. Μπορείτε να χρησιμοποιήσετε διαφορετικούς ταξινομητές ή να συνδυάσετε τα αποτελέσματα περισσότερων από έναν ταξινομητές. 

### Παράδοση Εργασίας

Η εργασία είναι είτε ατομική ή μπορεί να γίνει σε ομάδες το πολύ 3 ατόμων. Η τελική αξιολόγηση θα βασίζεται τόσο στο micro F1-score που θα επιτύχετε, όσο και στη συνολική προσέγγισή σας στο πρόβλημα. Στα πλαίσια της εργασίας, θα πρέπει να υποβληθούν τα εξής:
<ul>
    <li>Μια αναφορά 2 σελίδων, στην οποία θα πρέπει να περιγράψετε την προσέγγιση και τις μεθόδους που χρησιμοποιήσατε. Δεδομένου ότι πρόκειται για ένα πραγματικό πρόβλημα ταξινόμησης, μας ενδιαφέρει να γνωρίζουμε πώς αντιμετωπίσατε κάθε στάδιο του προβλήματος, π.χ. τι είδους αναπαράσταση δεδομένων χρησιμοποιήσατε, τι χαρακτηριστικά χρησιμοποιήσατε, εάν εφαρμόσατε τεχνικές μείωσης διάστασης, ποιούς αλγορίθμους ταξινόμησης δοκιμάσατε και γιατί, την απόδοση των μεθόδων σας (macro F1-score και χρόνο εκπαίδευσης), τυχόν προσεγγίσεις που τελικά δεν λειτούργησαν, αλλά
είναι ενδιαφέρον να παρουσιάσετε, και γενικά, ότι νομίζετε ότι είναι ενδιαφέρον να αναφερθεί.</li>
    <li>Ενα φάκελο με τον κώδικα της εφαρμογής σας.</li>
    <li>Εναλλακτικά μπορείτε να συνδυάσετε τα δυο παραπάνω σε ένα αρχείο Ipython Notebook.</li>
    <li>Δημιουργήστε ένα αρχείο .zip που περιέχει τον κώδικα και την αναφορά και υποβάλετέ τον στην πλατφόρμα e-class.</li>
    <li>**Λήξη προθεσμίας**: 6 Ιανουαρίου 2019.</li>
</ul>